zookeeper/parent_node_watcher.go

```diff
@@ -0,0 +1,260 @@
+package zookeeper
+
+import (
+	"time"
+
+	"github.com/jpillora/backoff"
+	"github.com/samuel/go-zookeeper/zk"
+	"github.com/sirupsen/logrus"
+)
+
+// ParentNodeWatchListener function signature for parent node watcher listner, this is used to invoke a callback when a ZK node changes
+type ParentNodeWatchListener func([]string)
+
+type parentNodeWatcher struct {
+	client       ZKClient
+	nodePath     string
+	pollTimeout  time.Duration
+	lastVersion  int32
+	children     []string
+	eventChannel <-chan zk.Event
+	disconnected chan struct{}
+	closed       chan struct{}
+	listener     ParentNodeWatchListener
+	watchActive  bool
+	log          *logrus.Entry
+}
+
+// CreateParentNodeWatcher returns a ZK ParentNodeWatcher that will call the provided listener when the children of the node are modified.
+func CreateParentNodeWatcher(client ZKClient, path string, polltimeout time.Duration, listener ParentNodeWatchListener) (ParentNodeWatcher, error) {
```

Comment:
This has no callers. What is this module used for?

Reply:
This is really my only blocking comment

Reply:
This will be used for resolving the update race condition. The leader of an update will be watching a zk node for master statuses which this Watcher will be used for.

I tested this in the zkVersionStore just by logging changes, but removed it as we don't need it there.